README.md
# Countdown text with animation.

### Preview:

![preview.gif](example/preview.gif)

### Feature:

- Auto calculate time pass.
- Build-in animations and custom animation builder.
- support reverse countDown.

### Basic Usage:

```dart
AnimateCountdownText(
               dateTime: DateTime(2020, 7, 16, 5, 13, 10),// Must provide
              format: (duration)=>DurationFormat(day:duration.inDays),
              characterTextStyle: TextStyle(fontSize: 16),
              animationType: AnimationType.bounceIn,
              reverse: true,
            )
```

### All parameters:

|Param|Desc|
|--|--|
|DateTime? dateTime|Compared DateTime, must provide.|
|FormatDuration format|Transform duration to human-readable format, (Duration)=>FormatDuration|
|AnimationType animationType|Build-in animation type, will ignore this if [animationBuilder] is provided|
|TextStyle characterTextStyle|Character TextStyle|
|TextStyle suffixTextStyle|Suffix TextStyle|
|Duration interval|Interval to refresh view, default to 1 second|
|bool reverse|Direction, default to false. Set it to true if you want to count reverse|
|Duration? expireDuration|The max duration that should mark countdown as done, If not null, [format] will return this as duration params when timeUp. Default to Duration.zero.|
|Duration? reverseExpireDuration|The max duration that should mark reverse countdown as done,<br/> If not null, [format] will return this as duration when timeUp. Default to null.|
|VoidCallback? onDone|Callback when timeUp. Only work when expireDuration is not null.|
|EdgeInsets characterPadding|Padding of characters, including character and suffix|
|AnimationBuilder? animationBuilder|Custom animation builder. |


CHANGELOG.md
## [1.0.0] - 2021/4/15.
 * initial release.
## [1.1.0] - 2021/4/23.
 * add param reverseExpireDuration.
 * fix bugs when update widget.
 * fix bugs in calculateYMDDistance extension.
 * update example.
## [1.1.1] - 2021/4/23.
 * fix bugs when update unit.
## [1.1.2] - 2021/7/20.
 * fix timeLeft calculate bug after phone wakeup.
 * remove initDuration param.
## [1.1.3] - 2021/7/20.
 * fix broadcastStream not canceled after disposed.


lib/extensions.dart
part of "animate_countdown_text.dart";

///
///@author shaw
///@date 2021/4/14
///@desc Extensions
///
extension DurationExt on Duration {
  /// Year Month Day in human words
  List<int> calculateYMDDistance({required bool reverse}) {
    DateTime now = DateTime.now().toUtc();
    DateTime to = reverse ? now.subtract(this) : now.add(this);
    DateTime start;
    DateTime end;
    bool negative = this.isNegative;

    if (negative) {
      start = to;
      end = now;
    } else {
      start = now;
      end = to;
    }

    var distYear = end.year - start.year;
    var distMonth = end.month - start.month;
    var distDay = end.day - start.day;
    if (distDay < 0) {
      distMonth--;
      if (distMonth < 0) {
        distYear--;
        distMonth = end.month + 12 - start.month - 1;
      }
      distDay = end.day + (start.daysOfMonth - start.day);
    } else if (distMonth < 0) {
      distYear--;
      distMonth = end.month + 12 - start.month;
    }
    return negative ? [-distYear, -distMonth, -distDay] : [distYear, distMonth, distDay];
  }

  /// Year Month Day Hour Minute Second in human words
  List<int> calculateYMDHMSDistance({required bool reverse}) {
    final ymdList = this.calculateYMDDistance(reverse: reverse);
    ymdList.add(this._hoursDistance);
    ymdList.add(this._minutesDistance);
    ymdList.add(this._secondsDistance);
    return ymdList;
  }

  /// Day Hour Minute Second
  List<int> calculateDHMSDistance() {
    int distDay = this.inDays;
    int disHour = this._hoursDistance;
    int disMinute = this._minutesDistance;
    int disSecond = this._secondsDistance;
    return [distDay, disHour, disMinute, disSecond];
  }

  int get _hoursDistance =>
      this.inHours.isNegative ? -(this.inHours.abs() % Duration.hoursPerDay) : this.inHours % Duration.hoursPerDay;

  int get _minutesDistance =>
      this.inMinutes.isNegative ? -(this.inMinutes.abs() % Duration.minutesPerHour) : this.inMinutes % Duration.minutesPerHour;

  int get _secondsDistance => this.inSeconds.isNegative
      ? -(this.inSeconds.abs() % Duration.secondsPerMinute)
      : this.inSeconds % Duration.secondsPerMinute;
}

extension on DateTime {
  /// return day count of month
  int get daysOfMonth {
    if (month == 2) {
      if (year % 4 == 0 && year % 100 != 0 || year % 400 == 0) {
        // leap
        return 29;
      }
      return 28;
    }
    if ([1, 3, 5, 7, 8, 10, 12].contains(month)) {
      return 31;
    }
    return 30;
  }
}


lib/animate_unit.dart
part of "animate_countdown_text.dart";

class AnimateUnit<String> extends StatefulWidget {
  final Stream<String?> itemStream;
  final String initValue;
  final TextStyle textStyle;
  final AnimationType animationType;
  final AnimationBuilder? animationBuilder;
  final EdgeInsets padding;

  AnimateUnit(
      {Key? key,
      required this.itemStream,
      required this.initValue,
      required this.textStyle,
      required this.animationType,
      required this.padding,
      this.animationBuilder})
      : super(key: key);

  @override
  _DigitState createState() => _DigitState();
}

class _DigitState extends State<AnimateUnit> with SingleTickerProviderStateMixin {
  // ignore: cancel_subscriptions
  StreamSubscription<String?>? _streamSubscription;
  late String? _currentValue;
  String? _preValue;
  final Duration _animationDuration = const Duration(milliseconds: 600);

  @override
  void initState() {
    _init();
    super.initState();
  }

  @override
  void didUpdateWidget(covariant AnimateUnit oldWidget) {
    if (oldWidget.initValue != widget.initValue) {
      _streamSubscription?.cancel();
      _init();
    }
    super.didUpdateWidget(oldWidget);
  }

  _init() {
    _currentValue = widget.initValue;
    _streamSubscription = widget.itemStream.distinct().listen((value) {
      _notifyValueChange(value);
    }) as StreamSubscription<String?>?;
  }

  _notifyValueChange(String? newValue) {
    if (mounted) {
      setState(() {
        _preValue = _currentValue;
        _currentValue = newValue;
      });
    }
  }

  @override
  Widget build(BuildContext context) {
    if (_currentValue == null) return SizedBox();

    final widgets = <Widget>[];
    final strList = _currentValue!.split('');
    for (int i = 0; i < strList.length; i++) {
      final character = strList[i];
      widgets.add(Container(
          padding: widget.padding,
          alignment: Alignment.center,
          child: _buildAnimation(child: Text(character, style: widget.textStyle), character: character, index: i)));
    }
    return Row(
      children: widgets,
    );
  }

  _buildAnimation({required Widget child, required String character, required int index}) {
    String preCharacter = "";
    if (_preValue != null && _preValue!.length >= index + 1) {
      preCharacter = _preValue![index];
    }

    if (widget.animationBuilder != null) {
      return widget.animationBuilder!(child, preCharacter, character);
    }

    switch (widget.animationType) {
      case AnimationType.fallDown:
        return Stack(
          children: [
            if (preCharacter.isNotEmpty && preCharacter != character)
              FadeOutDown(
                  key: ValueKey(preCharacter),
                  child: Text(preCharacter, style: widget.textStyle),
                  animate: true,
                  from: 10,
                  duration: const Duration(milliseconds: 200)),
            SlideInDown(key: ValueKey(character), child: child, from: 10, duration: _animationDuration)
          ],
        );
      case AnimationType.evaporation:
        return Stack(
          children: [
            if (preCharacter.isNotEmpty && preCharacter != character)
              FadeOutUp(
                  key: ValueKey(preCharacter),
                  child: Text(preCharacter, style: widget.textStyle),
                  from: 15,
                  animate: true,
                  duration: _animationDuration),
            FadeIn(key: ValueKey(character), duration: _animationDuration, child: child)
          ],
        );
      case AnimationType.scaleIn:
        return ZoomIn(key: ValueKey(character), child: child, duration: _animationDuration);
      case AnimationType.bounceIn:
        return BounceInUp(key: ValueKey(character), child: child, from: 10, duration: _animationDuration);
      case AnimationType.none:
        return child;
    }
  }

  @override
  void dispose() {
    _streamSubscription?.cancel();
    super.dispose();
  }
}


lib/duration_format.dart
part of 'animate_countdown_text.dart';

///
///@author shaw
///@date 2021/4/12
///@desc format of DateTime
///
class DurationFormat {
  const DurationFormat(
      {this.year,
      this.yearSuffix = "Year",
      this.month,
      this.monthSuffix = "Mon",
      this.day,
      this.daySuffix = "Day",
      this.hour,
      this.hourSuffix = ":",
      this.minute,
      this.minuteSuffix = ":",
      this.second,
      this.secondSuffix});

  final String? year;
  final String? yearSuffix;
  final String? month;
  final String? monthSuffix;
  final String? day;
  final String? daySuffix;
  final String? hour;
  final String? hourSuffix;
  final String? minute;
  final String? minuteSuffix;
  final String? second;
  final String? secondSuffix;

  bool get showYear => this.year?.isNotEmpty ?? false;

  bool get showYearSuffix => this.yearSuffix?.isNotEmpty ?? false;

  bool get showMonth => this.month?.isNotEmpty ?? false;

  bool get showMonthSuffix => this.monthSuffix?.isNotEmpty ?? false;

  bool get showDay => this.day?.isNotEmpty ?? false;

  bool get showDaySuffix => this.daySuffix?.isNotEmpty ?? false;

  bool get showHour => this.hour?.isNotEmpty ?? false;

  bool get showHourSuffix => this.hourSuffix?.isNotEmpty ?? false;

  bool get showMinute => this.minute?.isNotEmpty ?? false;

  bool get showMinuteSuffix => this.minuteSuffix?.isNotEmpty ?? false;

  bool get showSecond => this.second?.isNotEmpty ?? false;

  bool get showSecondSuffix => this.secondSuffix?.isNotEmpty ?? false;

  @override
  String toString() {
    return "${year ?? ''}${yearSuffix ?? ''}${month ?? ''}${monthSuffix ?? ''}${day ?? ''}${daySuffix ?? ''}${hour ?? ''}${hourSuffix ?? ''}${minute ?? ''}${minuteSuffix ?? ''}${second ?? ''}${secondSuffix ?? ''}";
  }

  bool sameFormatWith(DurationFormat? another) {
    if (another == null) return false;
    return this.showYear == another.showYear &&
        this.showYearSuffix == another.showYearSuffix &&
        this.showMonth == another.showMonth &&
        this.showMonthSuffix == another.showMonthSuffix &&
        this.showDay == another.showDay &&
        this.showDaySuffix == another.showDaySuffix &&
        this.showHour == another.showHour &&
        this.showHourSuffix == another.showHourSuffix &&
        this.showMinute == another.showMinute &&
        this.showMinuteSuffix == another.showMinuteSuffix &&
        this.showSecond == another.showSecond &&
        this.showSecondSuffix == another.showSecondSuffix;
  }
}


lib/animate_countdown_text.dart
library animate_countdown_text;

import 'dart:async';

import 'package:animate_do/animate_do.dart';
import 'package:flutter/material.dart';

part 'animate_unit.dart';

part 'animation_type.dart';

part "duration_format.dart";

part 'extensions.dart';

///
///@author shaw
///@date 2021/4/12
///@desc
///

typedef FormatDuration = DurationFormat Function(Duration duration);
typedef AnimationBuilder = Widget Function(Widget child, String? preCharacter, String character);

class AnimateCountdownText extends StatefulWidget {
  AnimateCountdownText(
      {Key? key,
      required this.dateTime,
      required this.format,
      this.animationType = AnimationType.evaporation,
      this.characterTextStyle = const TextStyle(),
      this.suffixTextStyle = const TextStyle(),
      this.interval = const Duration(seconds: 1),
      this.expireDuration = Duration.zero,
      this.reverseExpireDuration,
      this.onDone,
      this.characterPadding = const EdgeInsets.all(1),
      this.animationBuilder,
      this.reverse = false})
      : assert(!interval.isNegative && interval != Duration.zero, "Interval must positive"),
        super(key: key);

  /// DateTime that should be compared with
  final DateTime dateTime;

  /// Format Duration to [DurationFormat]
  final FormatDuration format;

  /// Build-in animation type, will ignore this if [animationBuilder] is provided
  final AnimationType animationType;

  /// Character TextStyle
  final TextStyle characterTextStyle;

  /// Suffix TextStyle
  final TextStyle suffixTextStyle;

  /// Interval to refresh view, default to 1 second
  final Duration interval;

  /// Direction
  /// true: countdown
  /// false: timing
  final bool reverse;

  /// The max duration that should mark countdown as done,
  /// If not null, [format] will return this as duration when timeUp.
  /// default to Duration.zero
  final Duration? expireDuration;

  /// The max duration that should mark reverse countdown as done,
  /// If not null, [format] will return this as duration when timeUp.
  /// default to null
  final Duration? reverseExpireDuration;

  /// Callback when duration == [expireDuration]/[reverseExpireDuration].
  final VoidCallback? onDone;

  /// Padding of characters
  final EdgeInsets characterPadding;

  /// Custom animation
  final AnimationBuilder? animationBuilder;

  @override
  _AnimateCountdownTextState createState() => _AnimateCountdownTextState();
}

class _AnimateCountdownTextState extends State<AnimateCountdownText> {
  late Duration timeLeft;
  late final Stream<DurationFormat> _durationFormatStream;
  late DurationFormat _initDurationFormat;
  late Duration? expireDuration;
  StreamSubscription? subscription;

  @override
  void initState() {
    super.initState();
    _durationFormatStream = Stream<DurationFormat>.periodic(widget.interval, (_) {
      _updateTimeLeft();
      _checkAndHandleTimeUp();
      return widget.format(timeLeft);
    }).asBroadcastStream(onCancel: (subscription) => subscription.cancel());
    _init();
    subscription = _durationFormatStream.listen((format) {
      if (!this._initDurationFormat.sameFormatWith(format)) _refreshInitDurationFormat(format);
    });
  }

  @override
  void didUpdateWidget(covariant AnimateCountdownText oldWidget) {
    if (oldWidget.dateTime != this.widget.dateTime ||
        oldWidget.animationType != this.widget.animationType ||
        oldWidget.interval != this.widget.interval ||
        oldWidget.expireDuration != this.widget.expireDuration ||
        oldWidget.reverseExpireDuration != this.widget.reverseExpireDuration ||
        oldWidget.reverse != this.widget.reverse) {
      _init();
    }
    super.didUpdateWidget(oldWidget);
  }

  void _init() {
    this.expireDuration = widget.reverse ? widget.reverseExpireDuration : widget.expireDuration;

    _updateTimeLeft();

    _checkAndHandleTimeUp();

    _initDurationFormat = widget.format(timeLeft);
  }

  _updateTimeLeft() {
    Duration duration;

    DateTime now = DateTime.now();
    if (widget.dateTime.isUtc) {
      now = now.toUtc();
    }
    duration = widget.reverse ? now.difference(widget.dateTime) : widget.dateTime.difference(now);

    timeLeft = duration;
  }

  _refreshInitDurationFormat(DurationFormat newDurationFormat) {
    if (!mounted) return;
    setState(() {
      this._initDurationFormat = newDurationFormat;
    });
  }

  _checkAndHandleTimeUp() {
    if (this.expireDuration != null) {
      if (widget.reverse && timeLeft >= this.expireDuration! ||
          !widget.reverse && timeLeft <= this.expireDuration!) {
        widget.onDone?.call();
        timeLeft = this.expireDuration!;
      }
    }
  }

  @override
  Widget build(BuildContext context) {
    List<Widget> row = [];
    // Year
    if (_initDurationFormat.showYear) {
      row.add(_buildAnimateUnit((event) => event?.year, _initDurationFormat.year!));
      if (_initDurationFormat.showYearSuffix) {
        row.add(_buildSuffixItem(_initDurationFormat.yearSuffix!));
      }
    }

    // Month
    if (_initDurationFormat.showMonth) {
      row.add(_buildAnimateUnit((event) => event?.month, _initDurationFormat.month!));
      if (_initDurationFormat.showMonthSuffix) {
        row.add(_buildSuffixItem(_initDurationFormat.monthSuffix!));
      }
    }

    // Day
    if (_initDurationFormat.showDay) {
      row.add(_buildAnimateUnit((event) => event?.day, _initDurationFormat.day!));
      if (_initDurationFormat.showDaySuffix) {
        row.add(_buildSuffixItem(_initDurationFormat.daySuffix!));
      }
    }

    // Hour
    if (_initDurationFormat.showHour) {
      row.add(_buildAnimateUnit((event) => event?.hour, _initDurationFormat.hour!));
      if (_initDurationFormat.showHourSuffix) {
        row.add(_buildSuffixItem(_initDurationFormat.hourSuffix!));
      }
    }

    // Minute
    if (_initDurationFormat.showMinute) {
      row.add(_buildAnimateUnit((event) => event?.minute, _initDurationFormat.minute!));
      if (_initDurationFormat.showMinuteSuffix) {
        row.add(_buildSuffixItem(_initDurationFormat.minuteSuffix!));
      }
    }
    // Second
    if (_initDurationFormat.showSecond) {
      row.add(_buildAnimateUnit((event) => event?.second, _initDurationFormat.second!));
      if (_initDurationFormat.showSecondSuffix) {
        row.add(_buildSuffixItem(_initDurationFormat.secondSuffix!));
      }
    }

    return Row(
      children: row,
      mainAxisSize: MainAxisSize.min,
    );
  }

  /// Animate content
  _buildAnimateUnit(String? Function(DurationFormat?) streamConvert, String initValue) {
    return AnimateUnit(
        itemStream: _durationFormatStream.map(streamConvert),
        initValue: initValue,
        textStyle: widget.characterTextStyle,
        animationType: widget.animationType,
        padding: widget.characterPadding,
        animationBuilder: widget.animationBuilder);
  }

  /// Suffix
  _buildSuffixItem(String suffix) {
    return Text(suffix, style: widget.suffixTextStyle);
  }

  @override
  void dispose() {
    subscription?.cancel();
    super.dispose();
  }
}


example/lib/main.dart
import 'package:animate_countdown_text/animate_countdown_text.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Demo Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, this.title}) : super(key: key);

  final String? title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  DateTime dateTime = DateTime.now().add(Duration(days: 5, hours: 2, minutes: 14));
  DateTime reverseDateTime = DateTime.now().subtract(Duration(days: 6, hours: 6, minutes: 26));

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title!),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          crossAxisAlignment: CrossAxisAlignment.center,
          children: <Widget>[
            Text("Evaporation", style: Theme.of(context).textTheme.headline6),
            AnimateCountdownText(
                dateTime: dateTime,
                format: _formatHMS,
                characterTextStyle: TextStyle(fontSize: 16),
                suffixTextStyle: TextStyle(fontSize: 14),
                animationType: AnimationType.evaporation),
            SizedBox(height: 30),
            Text("Scale In", style: Theme.of(context).textTheme.headline6),
            AnimateCountdownText(
                dateTime: dateTime,
                format: _formatHMS,
                characterTextStyle: TextStyle(fontSize: 16),
                animationType: AnimationType.scaleIn),
            SizedBox(height: 30),
            Text("Fall Down", style: Theme.of(context).textTheme.headline6),
            AnimateCountdownText(
                dateTime: dateTime,
                format: _formatHMS,
                characterTextStyle: TextStyle(fontSize: 16),
                animationType: AnimationType.fallDown),
            SizedBox(height: 30),
            Text("Bounce In & reverse", style: Theme.of(context).textTheme.headline6),
            AnimateCountdownText(
              dateTime: dateTime,
              format: _formatHMS,
              characterTextStyle: TextStyle(fontSize: 16),
              animationType: AnimationType.bounceIn,
              reverse: true,
            ),
            SizedBox(height: 30),
            Text("Day Hour Minute Second", style: Theme.of(context).textTheme.headline6),
            AnimateCountdownText(
                dateTime: dateTime,
                format: (duration) => _formatDHMS(duration, false),
                characterTextStyle: TextStyle(fontSize: 16),
                expireDuration: null,
                animationType: AnimationType.bounceIn),
            SizedBox(height: 30),
            Text("Year Month Day Hour Minute Second, Reverse", style: Theme.of(context).textTheme.bodyText2),
            AnimateCountdownText(
                dateTime: reverseDateTime,
                format: (duration) => _formatYMDHMS(duration, true),
                characterTextStyle: TextStyle(fontSize: 16),
                reverse: true,
                animationType: AnimationType.bounceIn),
            SizedBox(height: 30),
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceAround,
              children: [
                OutlinedButton(onPressed: _addOneHour, child: Text("Add")),
                OutlinedButton(onPressed: _subtractOneHour, child: Text("Subtract"))
              ],
            )
          ],
        ),
      ),
    );
  }

  _addOneHour() {
    setState(() {
      dateTime = dateTime.add(Duration(hours: 1));
      reverseDateTime = reverseDateTime.add(Duration(hours: 1));
    });
  }

  _subtractOneHour() {
    setState(() {
      dateTime = dateTime.subtract(Duration(hours: 1));
      reverseDateTime = reverseDateTime.subtract(Duration(hours: 1));
    });
  }

  DurationFormat _formatHMS(Duration duration) {
    final hours = duration.inHours;
    final minutes = duration.inMinutes - hours * 60;
    final seconds = duration.inSeconds - hours * 60 * 60 - minutes * 60;

    return DurationFormat(
        hour: "$hours",
        hourSuffix: " Hour ",
        minute: "$minutes",
        minuteSuffix: " Min ",
        second: "$seconds",
        secondSuffix: " Sec ");
  }

  DurationFormat _formatDHMS(Duration duration, bool reverse) {
    List<int> format = duration.calculateDHMSDistance();
    bool dayExist = format[0] != 0;
    bool hourExist = format[1] != 0;
    bool minuteExist = format[2] != 0;
    bool secondExist = format[3] != 0;
    return DurationFormat(
      day: dayExist ? "${format[0]}" : null,
      daySuffix: "Day",
      hour: hourExist ? "${format[1]}" : null,
      hourSuffix: "Hour",
      minute: minuteExist ? "${format[2]}" : null,
      minuteSuffix: "Min",
      second: secondExist ? "${format[3]}" : null,
      secondSuffix: "Sec",
    );
  }

  DurationFormat _formatYMDHMS(Duration duration, bool reverse) {
    List<int> format = duration.calculateYMDHMSDistance(reverse: reverse);
    bool yearExist = format[0] != 0;
    bool monthExist = format[1] != 0;
    bool dayExist = format[2] != 0;
    bool hourExist = format[3] != 0;
    bool minuteExist = format[4] != 0;
    bool secondExist = format[5] != 0;
    bool timeUp = !yearExist && !monthExist && !dayExist && !hourExist && !minuteExist && !secondExist;
    return DurationFormat(
      year: yearExist ? "${format[0]}" : null,
      yearSuffix: "Year",
      month: monthExist ? "${format[1]}" : null,
      monthSuffix: "Mon",
      day: dayExist ? "${format[2]}" : null,
      daySuffix: "Day",
      hour: hourExist ? "${format[3]}" : null,
      hourSuffix: "Hour",
      minute: minuteExist ? "${format[4]}" : null,
      minuteSuffix: "Min",
      second: timeUp
          ? "0"
          : secondExist
              ? "${format[5]}"
              : null,
      secondSuffix: "Sec",
    );
  }
}
